driver/ixsystems/common.py

```diff
@@ -381,10 +381,21 @@ def _delete_snapshot(self, name, volume_name):
                 self.configuration.ixsystems_dataset_path + '/' + volume_name),
             name)
         LOG.debug('_delete_snapshot urn : %s', request_urn)
+        try:
+            ret = self.handle.invoke_command(FreeNASServer.SELECT_COMMAND,
+                                             request_urn, None)
+            LOG.debug('_delete_snapshot select response : %s', json.dumps(ret))
+            if ret['status'] == 'error' and ret['code'] == 404:
+                LOG.info("Attempting delete Cinder volume %s snapshot %s, however it cannot be found on TrueNAS"
+                         % (volume_name, name))
+                LOG.info("Assume TrueNAS admin delete it manually, proceeding with snapshot delete action on cinder side")
+                return
```

Comment:
Should you also log anything that is not the expected result, not only 404?

Or does invoke_command handles it?

Reply:
Current ix driver has two level of error logging for each Truenas API call, first on http level inside invoke_command code：
https://github.com/iXsystems/cinder/blob/master/driver/ixsystems/freenasapi.py#L210
And then log on driver function level, like _delete_snapshot function above for code 404 error handling.
Maybe we can improve this, but I don't have any idea now, please comments.

Reply:
No worry, I checked, urllib.error.HTTPError implement rfc2616 which covers all http standard errors from 400+ upwards https://docs.python.org/3/library/urllib.error.html  https://datatracker.ietf.org/doc/html/rfc2616.html#section-10.4 We should be good then?